### PR TITLE
Docs for extensions

### PR DESCRIPTION
**Pull request by mike0sv**
This is second PR for #171 (after #172); Also rel. #181
It adds separate page for each MLEM extension. For now most of them consist only of auto-generated content:
1. short description taken from module docstring (see https://github.com/iterative/mlem/pull/413)
2. (optional) requirements 
3. Implementation reference - generated reference for each MLEM class implemented in this extension

There are also two sections that are mostly empty: description and examples. They are meant to be filled manually. 
However I filled them for first pages in each category (model/data/etc) just as an example.

Open questions:

- [ ] Where to put this. For now it resides in a separate sidebar section. There is also an option to merge it with `User Guide`
- [ ] Formatting. I did my best, but we def can do better :). Please take a look at multiple pages and give your suggestions. Preferably use first page (`models/sklearn`) to leave comments and change suggestions so we have them all in one place. I will adapt bootstrap script and re-generate all of the pages, no need to leave them everywhere. 
- [ ] Actual content. As @shcheklein says, the actual work is writing those manual sections. Once we are finished with first two Qs, we can start on this one. Since it's a lot of work, I suggest we do each extension in a separate PR. Also no need to wait for all of them - we can just hide unfinished ones (just from sidebar or actually remove the files)

---

In review app: https://mlem-ai-feature-0-3-0-e-qpg1qt.herokuapp.com/doc/extensions

### CODE REVIEW

**Comment by shcheklein** [comment]
> Actual content. As @shcheklein says, the actual work is writing those manual sections. Once we are finished with first two Qs, we can start on this one. Since it's a lot of work, I suggest we do each extension in a separate PR. Also no need to wait for all of them - we can just hide unfinished ones (just from sidebar or actually remove the files)

sounds good to me @mike0sv !

**Comment by aguschin** [comment]
> Where to put this. For now it resides in a separate sidebar section. There is also an option to merge it with User Guide

Let's first see what extensions we have and what we want to write as docs:

1. Data formats  (to `save` and `import`) - numpy, pandas, etc  (split `save` and `import` maybe?)
2. ML frameworks  (to `save` and `import`) - onnx, tensorflow, etc
3. Ways to `serve` - fastapi, rabbitmq, etc
4. Artifacts to `build` - docker, pip package, onnx again, etc
5. Places to `deploy` to - heroku, k8s, etc
6. Git providers to work with - GH, GL, BB

I assume, eventually we would like to cover all of these. For the beginning, "deploy" and "build" probably gets the priority.

In docs, these items could be subpages of User Guide, e.g.: `User Guide > Deployment > Heroku` or `User Guide > ML frameworks > Onnx`.

**Comment by aguschin** [comment]
Ahah, I was writing the previous comment without checking out [the docs deployment to preview env](https://mlem-ai-feature-0-3-0-e-qpg1qt.herokuapp.com/doc/extensions/).

I wonder if we can merge some pages into what you generate. E.g. ["Working with data"](https://mlem-ai-feature-0-3-0-e-qpg1qt.herokuapp.com/doc/user-guide/data) could be moved here ["Data"](https://mlem-ai-feature-0-3-0-e-qpg1qt.herokuapp.com/doc/extensions/data). Looks like there are not so many opportunities though, cause Models/Deployemnts/Serve/Build was already covered in GS. So we can just add a link to GS there.

File: content/docs/extensions/model/sklearn.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+### Saving and loading Scikit-Learn model
+
+```python
+from sklearn.datasets import load_iris
+from sklearn.ensemble import RandomForestClassifier
+
+from mlem.api import save, load
+
+
+data, y = load_iris(return_X_y=True, as_frame=True)
```
Comment (aguschin) [inline content/docs/extensions/model/sklearn.md:29]: I wish we could auto-generate this from our tests, but not sure it's possible. We have asserts, fixtures, etc, it's going to be a challenge to remove them and make the code both working and readable.

File: content/docs/extensions/model/sklearn.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    rf,
+    "rf",
+    sample_data=data,
+)
+
+rf = load("rf")
+rf.predict(data)
+```
+
+## Implementation reference
```
[the remark this answers is not in the document]
Comment (aguschin) [reply content/docs/extensions/model/sklearn.md:43]: So if this is for experienced developers, I think we should keep this apart from the other part of User Guide. Maybe `UG > Extensions > Data > Numpy` or just `Extensions > Data > Numpy` as you have it now. And maybe move "Extending" and "MLEM ABCs" there as well.

**Comment by mike0sv** [comment]
Had a call with @aguschin, we decided to go this way:

Target docs section for specific user groups. There are 4 of them:
1. Novice - first time mlemming. His section is Get Started of course. It should showcase mlem out-of-the-box features without additional configuration
2. Average user - uses only out-of-the-box features, but extensively. User Guide is for him - it should explain mlem concepts in detail, but without internals. Basically all mlem out-of-the-box features with every possible additional configuration (e.g how to use mlem vs sagemaker), but again without implementation details - just everything needed to use it
3. Advanced user - somebody who is in need of custom implementations and should know internal details. We need new section for those, eg "Advanced Usage"
4. Contributor - the same as 3. plus contributing guide

So, we will merge `Extensions` section into `User Guide`: each `extensions/<type>` page will become `user-guide/how to do <feature type> in mlem` (not exact wording of course) and each `extensions/<type>/<subtype>` will become a subpage `how to do <type> with <subtype>`. So rn extension pages are more like `mlem.contrib.<ext>` python module description, and they will become about "how to use mlem with <ext>".

We also will re-write some User Guide pages to remove internal details from them. Or completely move some pages to "Advanced Usage" (I'm looking at you, `mlem abcs` and `extending`). Also, `implementation reference` sections of extension pages will be cut from them and moved into new "Objects Reference" page (maybe it will be top section like api and command reference, maybe under "Advanced Usage")

Does this make sense?

**Comment by aguschin** [comment]
@jorgeorpinel @shcheklein WDYT about what Mike said?

**Comment by shcheklein** [comment]
Yep, sounds very reasonable to me, folks.

File: content/docs/sidebar.json [hunk 3]
```diff
@@ -126,6 +126,167 @@
       }
     ]
   },
+  {
+    "slug": "extensions",
+    "label": "Extensions",
```
Comment (jorgeorpinel) [inline content/docs/sidebar.json:131]: Should probably be a section of the User Guide? Also no need for this label (slug is auto-capitalized)
```suggestion
    "slug": "extensions",
```

Comment (jorgeorpinel) [reply content/docs/sidebar.json:131]: Actually the pages generated look like references ([e.g.](https://mlem-ai-feature-0-3-0-e-qpg1qt.herokuapp.com/doc/extensions/model/sklearn)), it could be called **Extension Reference** (listed under **Command Reference**).

File: content/docs/extensions/model/index.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Model extensions
+
+Model extensions add support for new types models that MLEM can covert into MLEM
+model objects in [`save` API method](/doc/api-reference/save)
```
Comment (jorgeorpinel) [inline content/docs/extensions/model/index.md:4]: We should try to explain this at a higher level. What functionality do model extensions give to users? OK they extend MLEM model objects I think, how is that useful? Should we start with a recap of what a model object is (used for) first?

File: content/docs/extensions/model/index.md [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+
+Typicaly they will implement [ModelType](/doc/user-guide/mlem-abcs#modeltype)
+and [ModelIO](/doc/user-guide/mlem-abcs#modelio) interfaces.
+
+Some also implement [DataType](/doc/user-guide/mlem-abcs#datatype) interface if
+specific data objects are needed for model to work.
```
Comment (jorgeorpinel) [inline content/docs/extensions/model/index.md:10]: For now just a copy edit summarizing this part but let's consider (or make an issue about) auto-linking many of these ABCs as well as any other reference-like terms (commands, extensions even, other types, etc).

```suggestion
Typically they will implement [ModelType] and [ModelIO] interfaces. Some also
implement the [DataType] interface if specific MLEM data objects are needed for
model to work.

[modeltype]: /doc/user-guide/mlem-abcs#modeltype
[modelio]: /doc/user-guide/mlem-abcs#modelio
[datatype]: /doc/user-guide/mlem-abcs#datatype
```

**Review by jorgeorpinel — COMMENTED** [review]
> the actual work is writing those manual sections

I'm just going to leave a couple comments on one of the index pages for now (probably apply to all):

File: content/docs/sidebar.json (hunk 3 above)
Comment (jorgeorpinel) [reply content/docs/sidebar.json:131]: > In docs, these items could be subpages of User Guide, e.g.: User Guide > Deployment > Heroku or User Guide > ML frameworks > Onnx.

@aguschin this seems to make sense but the type of content is definitely very much a technical reference so I think it's easier to make a separate section for these extension references.

_BTW I also think current parts of the User Guide like [MLEM ABCs](https://mlem.ai/doc/user-guide/mlem-abcs) are basically references and should be moved/split to their own sections as well._

We can and should still have guides about these topics but without all the specs and formalities, and link to the refs as needed instead. Divide and conquer ⚔️

> merge some pages into what you generate. E.g. ["Working with data"](https://mlem-ai-feature-0-3-0-e-qpg1qt.herokuapp.com/doc/user-guide/data) could be moved here ["Data"](https://mlem-ai-feature-0-3-0-e-qpg1qt.herokuapp.com/doc/extensions/data)

Again this sounds reasonable and if that's the only way to manage things for now (mixing types of documents in different sections) OK but ideally a clear separation between content types goes better with our [docs framework](https://www.notion.so/iterative/Framework-7e86e484e50e42529df3dbfe32b4e987).

File: content/docs/extensions/model/sklearn.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
+# Scikit-Learn Models Support
+
+[ModelType](/doc/user-guide/mlem-abcs#modeltype) implementations for any
+sklearn-compatible classes as well as `Pipeline`
```
Comment (jorgeorpinel) [inline content/docs/extensions/model/sklearn.md:4]: The URLs and titles aren't very consistent: `sklearn` vs. **Scikit-Learn Models Support**. Consistency reduces confusion and can help SEO.

Product questions (rethorical i.e. these need links or better explanations IMO):
- What are sklearn-compatible classes ?
- What is `Pipeline`?

However, this intro is a bit low-level (even for a reference). It's OK to be specific about what the extension is/does but can we try to zoom out, use more general language, etc? Something like "Enables you to connect Scikit-Learn classes to MLEM by implementing a model object..." (just an idea)

File: content/docs/extensions/model/sklearn.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
+# Scikit-Learn Models Support
```
Comment (jorgeorpinel) [inline content/docs/extensions/model/sklearn.md:1]: The URLs and titles aren't very consistent: `sklearn` vs. **Scikit-Learn Models Support**. Consistency reduces confusion and can help SEO.

File: content/docs/extensions/model/sklearn.md (hunk 6 above)
Comment (jorgeorpinel) [inline content/docs/extensions/model/sklearn.md:4]: Product questions (rhetorical i.e. these need links or better explanations IMO):
- What are sklearn-compatible classes ?
- What is `Pipeline`?

However, this intro is a bit low-level (even for a reference). It's OK to be specific about what the extension is/does but can we try to zoom out, use more general language, etc? Something like "Enables you to connect Scikit-Learn classes to MLEM by implementing a model object..." (just an idea)

File: content/docs/extensions/model/sklearn.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+[ModelType](/doc/user-guide/mlem-abcs#modeltype) implementations for any
+sklearn-compatible classes as well as `Pipeline`
+
+## Description
+
+**TODO**
```
Comment (jorgeorpinel) [inline content/docs/extensions/model/sklearn.md:8]: Haven't seen any descriptions yet. Any ideas for what will be in these sections? Maybe leave it for the user guide if it's unclear now.

File: content/docs/extensions/model/sklearn.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+## Requirements
+
+```bash
+pip install mlem[sklearn]
+# or
+pip install scikit-learn
+```
```
Comment (jorgeorpinel) [inline content/docs/extensions/model/sklearn.md:16]: - **Installation**?
- `bash`->`cli`?

File: content/docs/extensions/model/sklearn.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+## Examples
+
+### Saving and loading Scikit-Learn model
```
Comment (jorgeorpinel) [inline content/docs/extensions/model/sklearn.md:20]: Better to use `## Example: Title` if you want them to get listed in the right-hand ToC automatically:

image

File: content/docs/extensions/model/sklearn.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jorgeorpinel) [reply content/docs/extensions/model/sklearn.md:43]: **Implementation reference** -> just **Classes** and/or **methods**? It's already clear we're in a reference doc.

But if this only repeats docstrings, IDE insight/completion I'm not sure we need to publish it here too (being discussed in #171).

**Review by jorgeorpinel — COMMENTED** [review]
> use first page (models/sklearn) to leave comments and change suggestions

👇🏼

**Comment by jorgeorpinel** [comment]
> Target docs section for specific user groups. There are 4 of them:

Thanks @mike0sv and @aguschin, great analysis. I've had the chance to think and research on this topic quite a bit and my view is somewhat different although the logic is similar. PTAL at this [proposed framework](https://www.notion.so/iterative/Framework-7e86e484e50e42529df3dbfe32b4e987#cc1d6d4ea1c2408d9892d7954aa98b12). However that conversation goes beyond the scope of this PR (mainly a class reference IMO) so I'll post a hidden answer down here but let's move it elsewhere after that?

<details>

In the proposed docs framework, we work with content levels and expected traffic (along with other properties) instead of audiences (but again, it's similar though). _These levels also roughly match marketing/SEO parts of the user journey (TOFU, MOFU, BOFU)_.

Some key comparisons:

- An audience you're missing is explorers who are finding for solutions. They look at very high level materials e.g. use cases and may or may not be technical enough or have time to look at anything else. We then hope the pass it on to others in the team to actually try, learn, adopt, etc.
- Novice: GS is the most important section for what the framework calls _adoption_ indeed. But it doesn't have to limit to the absolute simplest no-config usage: it's goal is to get the new user to the product value ASAP.
- Average: I think more or recurrent users, and they tend to use CLI or API (IDE) help a lot or the technical references the most. They may read parts of the UG now and then (when going from novice to avg or from avg to advanced for example).
- Advanced: I don't think there's need for special advanced usage sections. All docs (esp. mid/low-level) tend to be structured from basic/overview to detailed/advanced.
- Contributor: They usually get to the README first but having a contributors guide is def. nice.

In the end we have to decide on a view so there's no precise right or wrong, and a lot of this is biased towards DVC. So we should definitely be flexible and adapt the framework to also accommodate MLEM product and docs needs.

</details>

**Comment by jorgeorpinel** [comment]
> rn extension pages are more like mlem.contrib.<ext> python module description, and they will become about "how to use mlem with ".

I see the auto-generated references as very far from this plan. "How to" guides could be a great addition but it seems like we still need the more technical references and these are different tasks so the guides should be a different PR I think. Also, I'd make a single guide per object category and use [tabs](https://github.com/iterative/dvc.org/issues/2939) to consolidate all type into a single page.

File: content/docs/sidebar.json (hunk 3 above)
Comment (jorgeorpinel) [reply content/docs/sidebar.json:131]: > re-write some User Guide pages to remove internal details from them...
looking at you, mlem abcs and extending...
moved into new "Objects Reference" page

I do align with these ideas in general @mike0sv. Like I said, separate formalities into reference sections when possible 🙂

Comment (aguschin) [reply content/docs/sidebar.json:131]: Ok @jorgeorpinel, looks like we all agree on the summary @mike0sv posted. Some comments of mine above are outdated (written before the summary), so let me just resolve them.

File: content/docs/extensions/model/sklearn.md (hunk 7 above)
Comment (jorgeorpinel) [reply content/docs/extensions/model/sklearn.md:1]: Up to you on this. Just keep in mind URL vs page titles please 🙂 If they're different it should be a reasoned, conscious decision, I think.

**Comment by aguschin** [comment]
@jorgeorpinel @mike0sv, I'm closing this PR, since we decided that everything should be done in #188. @jorgeorpinel, I'll use your feedback above for #188.

File: content/docs/extensions/model/index.md (hunk 4 above)
Comment (aguschin) [reply content/docs/extensions/model/index.md:4]: I tried to address this in #188 in "Object Reference" section, e.g. see this page https://mlem-ai-new-docs-struct-hy95mv.herokuapp.com/doc/object-reference/model. If you have some feedback, you can post it here or do it in #188.

Comment (jorgeorpinel) [reply content/docs/extensions/model/index.md:4]: Unable to leave feedback at this level in #188 atm, it's too huge.
